File: studio/repositories/conversations_repo.py
"""Conversation turn persistence."""

import time

from studio import db
from studio.models import Conversation, to_doc


def ensure(project_id):
    with db.session() as s:
        row = s.query(Conversation).filter_by(projectId=project_id).first()
        if row:
            return to_doc(row)
        row = Conversation(projectId=project_id, turns=[])
        s.add(row)
        s.commit()
        return to_doc(row)


def append_turn(project_id, role, text, meta=None):
    turn = {
        "role": role,
        "text": text,
        "ts": time.time(),
        "meta": meta or {},
    }
    with db.session() as s:
        row = s.query(Conversation).filter_by(projectId=project_id).first()
        if not row:
            row = Conversation(projectId=project_id, turns=[])
            s.add(row)
            s.flush()
        turns = list(row.turns or [])
        turns.append(turn)
        row.turns = turns
        s.commit()
    return turn


def list_turns(project_id):
    doc = ensure(project_id)
    return list(doc.get("turns", []))

File: src/server/repos/conversationsRepo.ts
import { Prisma } from "@/generated/prisma/client";
import { prisma } from "../db";
import { toDoc } from "./doc";

export type Turn = {
  role: string;
  text: string;
  ts: number;
  meta?: Record<string, unknown>;
};

export async function ensure(projectId: string): Promise<Record<string, unknown>> {
  let row = await prisma().conversation.findUnique({ where: { projectId } });
  if (!row) {
    row = await prisma().conversation.create({ data: { projectId, turns: [] } });
  }
  return (toDoc(row) ?? {}) as Record<string, unknown>;
}

export async function appendTurn(
  projectId: string,
  role: string,
  text: string,
  meta?: Record<string, unknown>,
): Promise<Turn> {
  const turn: Turn = { role, text, ts: Date.now() / 1000, meta: meta ?? {} };
  const row = await prisma().conversation.findUnique({ where: { projectId } });
  const turns = [...((row?.turns as Turn[] | null) ?? []), turn] as Prisma.InputJsonValue;
  if (row) {
    await prisma().conversation.update({ where: { projectId }, data: { turns } });
  } else {
    await prisma().conversation.create({ data: { projectId, turns } });
  }
  return turn;
}

export async function listTurns(projectId: string): Promise<Turn[]> {
  const doc = await ensure(projectId);
  return [...((doc.turns as Turn[] | undefined) ?? [])];
}
